src/tracing/on_disk.rs

```diff
@@ -0,0 +1,95 @@
+use std::collections::HashSet;
+use std::fs;
+
+use anyhow::Context as _;
+use serde::{Deserialize, Serialize};
+use tracing_subscriber::prelude::*;
+use tracing_subscriber::{filter, fmt, registry};
+
+use super::*;
+
+#[derive(Clone, Debug, Default, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(default)]
+pub struct Config {
+    pub enabled: Option<bool>,
+    pub log_file: Option<String>,
+    pub log_level: Option<String>,
+    pub span_events: Option<HashSet<config::SpanEvent>>,
+}
+
+impl Config {
+    pub fn merge(&mut self, other: Self) {
+        self.enabled = other.enabled.or(self.enabled.take());
+        self.log_file = other.log_file.or(self.log_file.take());
+        self.log_level = other.log_level.or(self.log_level.take());
+        self.span_events = other.span_events.or(self.span_events.take());
+    }
+}
+
+#[rustfmt::skip] // `rustfmt` formats this into unreadable single line :/
+pub type Logger<S> = filter::Filtered<
+    Option<Layer<S>>,
+    filter::EnvFilter,
+    S,
+>;
+
+#[rustfmt::skip] // `rustfmt` formats this into unreadable single line :/
+pub type Layer<S> = fmt::Layer<
+    S,
+    fmt::format::DefaultFields,
+    fmt::format::Format,
+    fs::File,
+>;
+
+pub fn new_logger<S>(config: &mut Config) -> Logger<S>
+where
+    S: tracing::Subscriber + for<'span> registry::LookupSpan<'span>,
+{
+    let layer = match new_layer(config) {
+        Ok(layer) => layer,
+        Err(err) => {
+            eprintln!(
+                "failed to enable logging into {} log-file: {err}",
+                config.log_file.as_deref().unwrap_or(""),
+            );
+
+            config.enabled = Some(false);
+            None
+        }
+    };
+
+    let filter = new_filter(config);
+    layer.with_filter(filter)
+}
+
+pub fn new_layer<S>(config: &Config) -> anyhow::Result<Option<Layer<S>>>
+where
+    S: tracing::Subscriber + for<'span> registry::LookupSpan<'span>,
+{
+    if !config.enabled.unwrap_or_default() {
+        return Ok(None);
+    }
+
+    let Some(log_file) = &config.log_file else {
+        return Ok(None);
+    };
+
+    let writer = fs::OpenOptions::new()
+        .create(true)
+        .append(true)
+        .open(log_file)
+        .with_context(|| format!("failed to open {} log-file", log_file))?;
+
+    let layer = fmt::Layer::default()
+        .with_writer(writer)
+        .with_span_events(config::SpanEvent::unwrap_or_default_config(
+            &config.span_events,
+        ))
+        .with_ansi(false);
```

Comment:
As far as I can see, no buffering happens. It seems to create a writer directly, on which it writes directly.

Maybe we can wrap this writer in a [`BufWriter`](https://doc.rust-lang.org/std/io/struct.BufWriter.html)?

Reply:
Done. Had to wrap `io::BufWriter<fs::File>` with a `Mutex` to do that, though. `&fs::File` implements `io::Write`, but any other custom `io::Write` requires `&mut`, of course.

Reply:
But I guess `fs::File` (somehow) serializes writes anyway, so synchronization should not tank performance, I guess?

Reply:
I'm confused on why the buffered writer needs the mutex, and the file does not.

Reply:
`io::Write` requires `&mut`. `fs::File` is the exception, because `io::Write` is (somehow) implemented for `&fs::File`.

Just imagine, what would happen if you try to write into `io::BufWriter` from two different threads. How do you synchronize access to the internal buffer?

Reply:
> fs::File is the exception, because io::Write is (somehow) implemented for &fs::File.

I see, and you say you expect the syscalls to be serialized by the kernel.

Reply:
:heavy_check_mark:

Reply:
I have no idea how it works internally (probably on the OS/kernel level, as you say), tbh. 😅

But I do know that `&fs::File` implements `io::Write` and that tracing writes data correctly (e.g., without garbling it between different log lines) when using plain `fs::File`.

Reply:
🪄🔮